### ISSUE TITLE
Add metadata support for derivations

### ISSUE THREAD

**Issue by dkjii-g** [NixOS/nix#5327]
**Is your feature request related to a problem? Please describe.**
Nix does not have a way to have metadata on files. This causes issues with PaX (https://github.com/NixOS/nix/issues/185), selinux (https://github.com/NixOS/rfcs/pull/41#issuecomment-543777966), setuid binaries (https://github.com/NixOS/nix/issues/5177, even though this one is for building, https://github.com/NixOS/nixpkgs/issues/54401 for runtime, and making this work requires wrappers when used at runtime) and other xattrs such as capabilities I assume (haven't seen much about this).

**Describe the solution you'd like**
Add a privileged post_inst hook like gentoo where xattrs can be added to the final package. This way privileged users could install derivations which use that hook to set important flags such as setuid, security labels of PaX flags.

**Describe alternatives you've considered**
Not sure there is an alternative, apart from not supporting these features. Not everything can be done using wrappers.

**Additional context**
Metadata has become pretty important in for some features in Linux and imo ignoring it is a big drawback nix has right now.

I came upon this issue when I was looking into selinux nix support and this is the main stopping block I'm seeing in implementing it in a functional way (main issue in https://github.com/NixOS/rfcs/pull/41#issuecomment-543777966 afaik). Implementing it as an alias is possible but suboptimal IMO.

Might not be the right venture for the request, but I have not seen a good discussion about this in the issue tracker or discourse (it's pretty scattered).

**Comment by edolstra** [comment]
I don't think this will ever be implemented, because modifying files in the Nix store after they have been built to add setuid flags etc. violates immutability, in a very observable way (a setuid-root program behaves differently than a regular program). It would also be bad to have old setuid binaries laying around in the Nix store. This is why setuid support in NixOS is implemented via wrappers outside of the Nix store.

Regarding SELinux, I think AppArmor is more compatible with Nix because it doesn't require attaching metadata to files.

**Comment by dkjii-g** [comment]
> because modifying files in the Nix store after they have been built to add setuid flags etc. violates immutability, in a very observable way

I used post_inst as an example because I don't think gentoo has a postBuild hook running as root (might be wrong).

An alternative, functional and immutable, way to do it would be to have something that's executed as root before the package is finished installing into $out. Similar to what is postInstall (I think?). For example, iputils could be done in such a way:

```
[...]
privPostInstall = ''
  setcap cap_net_raw+p $out/bin/ping
''
```

This way the expression deterministically determines what's in the final $out and nothing outside of $out is touched.

>It would also be bad to have old setuid binaries laying around in the Nix store.

This might be an issue, but perhaps it's better fixed with having a CVE scanner or forcing gc on packages having such a hook.

**Comment by spacekitteh** [comment]
>I don't think this will ever be implemented, because modifying files in the Nix store after they have been built to add setuid flags etc. violates immutability,

Are you referring to the hashing not taking it into account? If so, couldn't the hashing algorithm be extended to take it into account?

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
